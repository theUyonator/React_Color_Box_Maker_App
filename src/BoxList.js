import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import { v4 as uuid } from "uuid";


const BoxList = () => {
  const [boxes, setBoxes] = useState([]);

  const remove = (id) => {
      setBoxes(boxes.filter(box => box.id !== id));
  };

  const renderBoxes = () => {
    return (
        <>
        {boxes.map(box => (
          <div key={box.id}>
            <Box width={box.width} height={box.height} backgroundColor={box.backgroundColor} remove={evt => remove(box.id)} />
          </div>
        ))}
        </>
    
    );
  };
  // end renderBoxes

  /** Add new box object to cart. */
  const addBox = box => {
    let newBox = { ...box, id: uuid() };
    setBoxes(boxes => [...boxes, newBox]);
  };
  // end addBox

  return (
    <div className="BoxList">
      <NewBoxForm addBox={addBox} />
      {renderBoxes()}
    </div>
  );
};
// end

export default BoxList;
