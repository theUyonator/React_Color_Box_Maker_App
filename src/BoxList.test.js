import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

describe("<BoxList /> rendering", () => {

it("renders correctly without crashing", () => {
    render(<BoxList />)
});

it("matches snapshot", () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

});

describe("Adds a new box and can remove a specific box", () => {
    it("can add a new box", function() {
        const { getByLabelText, queryByText } = render(<BoxList />);
      
        // no boxes yet
        expect(queryByText("X")).not.toBeInTheDocument();
      
        const widthInput = getByLabelText("Width:");
        const heightInput = getByLabelText("Height:");
        const backgroundcolorInput = getByLabelText("Background Color:");
        const submitBtn = queryByText("Add a new box!")
      
        // fill out the form
        fireEvent.change(widthInput, { target: { value: 6 }});
        fireEvent.change(heightInput, { target: { value: 6 }});
        fireEvent.change(backgroundcolorInput, { target: { value: "lilac" }});
        fireEvent.click(submitBtn);
      
        // box exists!
        // expect to see a box
        const removeButton = queryByText("X");
        expect(removeButton).toBeInTheDocument();
        expect(removeButton.previousSibling).toHaveStyle(`
        width: 6em;
        height: 6em;
        background-color: lilac;
      `);
      });

    it("can remove a box", function(){

        const { getByLabelText, queryByText } = render(<BoxList />);
      
        const widthInput = getByLabelText("Width:");
        const heightInput = getByLabelText("Height:");
        const backgroundcolorInput = getByLabelText("Background Color:");
        const submitBtn = queryByText("Add a new box!")
      
        // fill out the form
        fireEvent.change(widthInput, { target: { value: 6 }});
        fireEvent.change(heightInput, { target: { value: 6 }});
        fireEvent.change(backgroundcolorInput, { target: { value: "lilac" }});
        fireEvent.click(submitBtn);

        // expect to see a box
        const removeButton = queryByText("X");
        expect(removeButton).toBeInTheDocument();

        // Click the remove btn and expect it not to be in the document anymore
        fireEvent.click(removeButton);
        expect(removeButton).not.toBeInTheDocument();

    })
    
});

