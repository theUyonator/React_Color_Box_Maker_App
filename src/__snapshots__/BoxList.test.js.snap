// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoxList /> rendering matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form
      class="NewBoxForm"
    >
      <label
        class="NewBoxForm-label"
        for="width"
      >
        Width:
      </label>
      <input
        class="NewBoxForm-input"
        id="width"
        name="width"
        type="number"
        value="0"
      />
      <label
        class="NewBoxForm-label"
        for="height"
      >
        Height:
      </label>
      <input
        class="NewBoxForm-input"
        id="height"
        name="height"
        type="number"
        value="0"
      />
      <label
        class="NewBoxForm-label"
        for="backgroundColor"
      >
        Background Color:
      </label>
      <input
        class="NewBoxForm-input"
        id="backgroundColor"
        name="backgroundColor"
        type="text"
        value=""
      />
      <button
        class="NewBoxForm-btn"
        id="newBoxBtn"
      >
        Add a new box!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
